control/grpc.py

```diff
@@ -416,112 +438,781 @@ def delete_subsystem_safe(self, request, context=None):
+
+    def create_namespace(self, subsystem_nqn, bdev_name, nsid, anagrpid, uuid, context):
+        """Adds a namespace to a subsystem."""
+ 
+        if context:
+            assert self.omap_lock.locked()
+        nsid_msg = ""
+        if nsid and uuid:
+            nsid_msg = f" using NSID {nsid} and UUID {uuid}"
+        elif nsid:
+            nsid_msg = f" using NSID {nsid} "
+        elif uuid:
+            nsid_msg = f" using UUID {uuid} "
+
+        add_namespace_error_prefix = f"Failure adding namespace{nsid_msg}to {subsystem_nqn}"
+
+        self.logger.info(f"Received request to add {bdev_name} to {subsystem_nqn} with ANA group id {anagrpid}{nsid_msg}")
+
+        if anagrpid > MAX_ANA_GROUPS:
+            errmsg = f"{add_namespace_error_prefix}: Group ID {anagrpid} is bigger than configured maximum {MAX_ANA_GROUPS}"
+            self.logger.error(errmsg)
+            return pb2.nsid_status(status=errno.EINVAL, error_message=errmsg)
+
+        if self.is_discovery_nqn(subsystem_nqn):
+            errmsg = f"{add_namespace_error_prefix}: Can't add namespaces to a discovery subsystem"
+            self.logger.error(errmsg)
+            return pb2.nsid_status(status=errno.EINVAL, error_message=errmsg)
+
+        try:
+            nsid = rpc_nvmf.nvmf_subsystem_add_ns(
+                self.spdk_rpc_client,
+                nqn=subsystem_nqn,
+                bdev_name=bdev_name,
+                nsid=nsid,
+                anagrpid=anagrpid,
+                uuid=uuid,
+            )
+            self.logger.info(f"subsystem_add_ns: {nsid}")
+        except Exception as ex:
+            errmsg = f"{add_namespace_error_prefix}:\n{ex}"
+            self.logger.error(errmsg)
+            resp = self.parse_json_exeption(ex)
+            status = errno.EINVAL
+            if resp:
+                status = resp["code"]
+                errmsg = f"{add_namespace_error_prefix}: {resp['message']}"
+            return pb2.nsid_status(status=status, error_message=errmsg)
+
+        # Just in case SPDK failed with no exception
+        if not nsid:
+            self.logger.error(add_namespace_error_prefix)
+            return pb2.nsid_status(status=errno.EINVAL, error_message=add_namespace_error_prefix)
+
+        return pb2.nsid_status(nsid=nsid, status=0, error_message=os.strerror(0))
+
+    def find_unique_bdev_name(self) -> str:
+        try:
+            bdevs = rpc_bdev.bdev_get_bdevs(self.spdk_rpc_client)
+        except Exception as ex:
+            self.logger.error(f"Got exception while getting bdevs:\n{ex}")
+            bdevs = []
+        bdev_names = []
+        for bdev in bdevs:
+            try:
+                bdev_names.append(bdev["name"])
+            except Exception as ex:
+                self.logger.warning(f"Got exception while fteching bdev names:\n{ex}")
+                pass
+        random.seed()
+        randser = random.randint(1, 99999999999999)
```

Comment:
maybe we can construct a bdev name based on the namespace fields like uuid instead of using random numbers? would that approach result a unique name? another approach would be to enumerate bdevs, bdev_0, bdev_1, ..., bdev_N. WDYT?

Reply:
Using UUID is about the same as a random number as a random UUID will be created. I didn't bother with enumerating the bdevs as it it adds a state and more complication. As the chances we get an existing bdev name are very small I preferred to do it like that. In the worst case will create one or two extra bdev names. This is not really an issue and it saves us the complication of managing the bdev names.

Reply:
if using uuid to construct bdev name, like "bdev_$uuid" would result a unique bdev name, then probably we do not need to use random directly.

Reply:
It's not the UUID code has any better random number generator but fine, I'll change the code to use the UUID value for bedv name.